### PR TITLE
REAL working staging build!

### PR DESCRIPTION
**Pull request by Peedee2002**
Only functionality change:

https://github.com/csesoc/circles/assets/64952797/094ea8cb-6ac1-4768-95e7-bb1348857d4e

tiny fix that makes the degreeInfo get read properly, and to hide the menu if you selected the degree already. The degree  remains editable though.

This merges [!1061](https://github.com/csesoc/circles/pull/1061) as well.

This basically gets you a functional Circles user backend. Everything works, including graphical selector and progression checker.

Some tests are still failing, but tbh it is because those tests are broken.

The only slices still left are the courseTabs and settings slices, which I dont think we should remove tbh.

image

The backend tests pass tho

### CODE REVIEW

**Comment by Peedee2002** [comment]
I updated this with dev, and updated all dependencies to latest. I will then try to update it with code from:
- christian/feat/add-login-page
- react-query-prog-checker
- react-query-graphical-selector

For the other branches that are open rn @ollibowers  @martanman would you mind checking them and deleting all the ones that are either encompassed by this one (not the 3 listed there ofc) or letting me know what they are so i can add them here?

File: .vscode/settings.json [hunk 1]
```diff
@@ -2,25 +2,18 @@
...
     "python.testing.unittestEnabled": false,
     "python.testing.pytestEnabled": true,
     "python.testing.cwd": "${workspaceFolder}/backend",
-    "mypy.configFile": "./backend/mypy.ini",
     "files.exclude": {
         "**/__pycache__": true,
         "**/.hypothesis": true,
         "**/.pytest_cache": true
     },
-    "python.linting.pylintEnabled": true,
```
Comment (ollibowers) [inline .vscode/settings.json:14]: lol

File: frontend/src/components/GoogleAuthFlow/GoogleAuthFlow.tsx [hunk 2]
```diff
@@ -0,0 +1,30 @@
+import { useEffect } from 'react';
```
Comment (ollibowers) [inline frontend/src/components/GoogleAuthFlow/GoogleAuthFlow.tsx:1]: will probs need deleting later again, but is ok for now

File: frontend/src/utils/getAllCourseOfferings.ts [hunk 3]
```diff
@@ -53,4 +54,9 @@ export const courseHasOffering = (course: PlannerCourse, year: string, term: Ter
   return year in course.legacyOfferings && course.legacyOfferings[year].includes(term);
 };
 
+export const courseHasOfferingNew = (course: Course, term: Term): boolean => {
+  // TODO: Work out how legacy offerings work now
+  return course.terms.includes(term);
```
Comment (ollibowers) [inline frontend/src/utils/getAllCourseOfferings.ts:59]: will need to check if this actually works (after merge)

File: frontend/src/utils/api/coursesApi.ts [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+/* eslint-disable import/prefer-default-export */
+export const getCoursesInfo = async (): Promise<Record<string, Course>> => {
+  const courses: Record<string, Course> = {};
+  const planner = await getUserPlanner();
+
+  planner.years
+    .flatMap((x) => Object.values(x))
+    .flat()
+    .concat(planner.unplanned)
+    .forEach(async (id) => {
```
Comment (ollibowers) [inline frontend/src/utils/api/coursesApi.ts:14]: i question this, (but will address post merge)

File: frontend/src/pages/ProgressionChecker/Dashboard/Dashboard.tsx [hunk 5]
```diff
@@ -37,13 +39,19 @@ const Dashboard = ({ isLoading, structure, totalUOC, freeElectivesUOC }: Props)
...
   });
+  const coursesQuery = useQuery('courses', getUserCourses);
+  const courses = coursesQuery.data || badCourses;
+  const degreeQuery = useQuery('degree', getUserDegree);
+  const degree = degreeQuery.data || badDegree;
+  const { programCode } = degree;
 
-  const { courses } = useSelector((state: RootState) => state.planner);
-  const { programCode, programName } = useSelector((state: RootState) => state.degree);
+  const programName = (useQuery('progam', fetchAllDegrees).data?.programs || { [programCode]: '' })[
```
Comment (ollibowers) [inline frontend/src/pages/ProgressionChecker/Dashboard/Dashboard.tsx:48]: progam 🗣️ 🗣️

File: frontend/src/pages/TermPlanner/ExportPlannerMenu/ExportPlannerMenu.tsx [hunk 6]
```diff
@@ -45,8 +21,6 @@ const ExportPlannerMenu = ({ plannerRef }: Props) => {
       exportComponentAsPNG(plannerRef, exportFields);
     } else if (format === 'jpg') {
       exportComponentAsJPEG(plannerRef, exportFields);
-    } else if (format === 'json') {
-      exportComponentAsJSON();
```
Comment (ollibowers) [inline frontend/src/pages/TermPlanner/ExportPlannerMenu/ExportPlannerMenu.tsx:49]: we might want to bring this back via a backend route later

File: frontend/src/pages/TermPlanner/OptionsHeader/OptionsHeader.tsx [hunk 7]
```diff
@@ -120,24 +123,22 @@ const OptionsHeader = ({ plannerRef }: Props) => {
...
+              overlayClassName="popconfirm-unplan"
             >
               <S.OptionButton>
                 <FaRegCalendarTimes style={iconStyles} />
               </S.OptionButton>
             </Popconfirm>
           </Tooltip>
         )}
-
-        {areYearsHidden && (
```
Comment (ollibowers) [inline frontend/src/pages/TermPlanner/OptionsHeader/OptionsHeader.tsx:131]: again, might want to bring this back in the future? kinda a nice feature

File: frontend/src/pages/TermPlanner/TermBox/TermBox.tsx [hunk 8]
```diff
@@ -17,39 +20,65 @@ const Droppable = React.lazy(() =>
...
+  const queryClient = useQueryClient();
 
-  const { isSummerEnabled, completedTerms, courses } = useSelector(
-    (state: RootState) => state.planner
-  );
-  const [totalUOC, setTotalUOC] = useState(0);
-  const dispatch = useDispatch();
-  const handleCompleteTerm = () => {
-    dispatch(toggleTermComplete(name));
+  const toggleLockTerm = async () => {
```
Comment (ollibowers) [inline frontend/src/pages/TermPlanner/TermBox/TermBox.tsx:44]: move this out

**Pull request merged**